dali/core/geom_vec_test.cu

```diff
@@ -0,0 +1,160 @@
+  EXPECT_EQ(v2.x, 0) << "vec should be zero-initialized by default";
+  vec<3> v3;
+  EXPECT_EQ(v3.x, 0) << "vec should be zero-initialized by default";
+  EXPECT_EQ(v3.y, 0) << "vec should be zero-initialized by default";
+  EXPECT_EQ(v3.z, 0) << "vec should be zero-initialized by default";
+  vec<4> v4;
+  EXPECT_EQ(v4[0], 0) << "vec should be zero-initialized by default";
+  EXPECT_EQ(v4[1], 0) << "vec should be zero-initialized by default";
+  EXPECT_EQ(v4[2], 0) << "vec should be zero-initialized by default";
+  EXPECT_EQ(v4[3], 0) << "vec should be zero-initialized by default";
+  vec<5> v5;
+  EXPECT_EQ(v5[0], 0) << "vec should be zero-initialized by default";
+  EXPECT_EQ(v5[1], 0) << "vec should be zero-initialized by default";
+  EXPECT_EQ(v5[2], 0) << "vec should be zero-initialized by default";
+  EXPECT_EQ(v5[3], 0) << "vec should be zero-initialized by default";
+  EXPECT_EQ(v5[4], 0) << "vec should be zero-initialized by default";
+}
+
+TEST(Vec, FieldConstruct) {
+  vec<1> v1 = { 1 };
+  EXPECT_EQ(v1.x, 1);
+  EXPECT_EQ(v1[0], 1);
+  vec<2> v2 = { 1, 2 };
+  EXPECT_EQ(v2.x, 1);
+  EXPECT_EQ(v2.y, 2);
+  EXPECT_EQ(v2[0], 1);
+  EXPECT_EQ(v2[1], 2);
+  vec<3> v3 = { 1, 2, 3 };
+  EXPECT_EQ(v3.x, 1);
+  EXPECT_EQ(v3.y, 2);
+  EXPECT_EQ(v3.z, 3);
+  EXPECT_EQ(v3[2], 3);
+  vec<4> v4 = { 1, 2, 3, 4 };
+  EXPECT_EQ(v4.x, 1);
+  EXPECT_EQ(v4.y, 2);
+  EXPECT_EQ(v4.z, 3);
+  EXPECT_EQ(v4.w, 4);
+  EXPECT_EQ(v4[0], 1);
+  EXPECT_EQ(v4[1], 2);
+  EXPECT_EQ(v4[2], 3);
+  EXPECT_EQ(v4[3], 4);
+  vec<5> v5 = { 1, 2, 3, 4, 5 };
+  EXPECT_EQ(v5[0], 1);
+  EXPECT_EQ(v5[1], 2);
+  EXPECT_EQ(v5[2], 3);
+  EXPECT_EQ(v5[3], 4);
+  EXPECT_EQ(v5[4], 5);
+}
+
+TEST(Vec, Equality) {
+  vec<3> a = { 1, 2, 3 };
+  vec<3, int> b = { 1, 2, 3 };
+  vec<3> c = { 1, 2, 4 };
+  EXPECT_TRUE(a == b);
+  EXPECT_FALSE(a == c);
+  EXPECT_FALSE(a != b);
+  EXPECT_TRUE(a != c);
+}
+
+TEST(Vec, Op) {
+  vec<3> a = { 1, 2, 3 }, b = { 4, 5, 6 };
+  EXPECT_EQ(a+b, (vec<3>{5, 7, 9}));
+  EXPECT_EQ(a-b, (vec<3>{-3, -3, -3}));
+  EXPECT_EQ(a*b, (vec<3>{4, 10, 18}));
+  b = { 3, 2, 1 };
+  EXPECT_EQ((a<b), (vec<3, bool>(true, false, false)));
+  EXPECT_EQ((b<a), (vec<3, bool>(false, false, true)));
+}
+
+DEVICE_TEST(Dev_Vec, Op, 1, 1) {
+  vec<3> a = { 1, 2, 3 }, b = { 4, 5, 6 };
+  DEV_EXPECT_EQ(a+b, (vec<3>{5, 7, 9}));
+  DEV_EXPECT_EQ(a-b, (vec<3>{-3, -3, -3}));
+  DEV_EXPECT_EQ(a*b, (vec<3>{4, 10, 18}));
+  b = { 3, 2, 1 };
+  DEV_EXPECT_EQ((a<b), (vec<3, bool>(true, false, false)));
+  DEV_EXPECT_EQ((b<a), (vec<3, bool>(false, false, true)));
+}
+
+TEST(Vec, Cast) {
+  vec<4> a = { 1.2f, 2.4f, 3.4f, -5.3f };
+  EXPECT_EQ(a.cast<int>(), cast<int>(a));
+  EXPECT_EQ(a.cast<int>(), (vec<4, int>(1, 2, 3, -5)));
+}
+
+TEST(Vec, Iteration) {
+  const size_t N = 3;
+  vec<N> v;
+  EXPECT_EQ(dali::size(v), N);
+  for (size_t i = 0; i < N; i++)
+    v[i] = i + 5;
+  for (auto &x : v) {
+    auto ofs = &x - &v[0];
+    ASSERT_TRUE(ofs >= 0 && ofs < static_cast<ptrdiff_t>(N));
```

Comment:
```suggestion
    ASSERT_TRUE(ofs >= 5 && ofs < static_cast<ptrdiff_t>(N));
```

Reply:
Nope, it's a _pointer differnce_ that's tested here.